### ISSUE TITLE
[BUG] GCC High client tokens Confidential Client is not supported in Cross Cloud request

### ISSUE THREAD

**Issue by id-idg** [pnp/powershell#189]
### Reporting an Issue or Missing Feature
Reporting an issue

### Expected behavior 
Connect to Sharepoint USGovernment tenant via app registration (GCC High) and update template
![image](https://user-images.githubusercontent.com/77301614/106046346-23703980-60b0-11eb-9966-404cbb7858e1.png)

### Actual behavior 

![image](https://user-images.githubusercontent.com/77301614/106046315-16534a80-60b0-11eb-909c-96a95795b70a.png)

### Steps to reproduce behavior
`Connect-PnPOnline -AzureEnvironment USGovernment -Url https://[TENANT].sharepoint.us/sites/benefits/ -ClientId [id] -ClientSecret [secret]`

`Apply-PnPProvisioningTemplate -path .\benefits\benefits.pnp`

### What is the version of the Cmdlet module you are running?
1.2.0

### Which operating system/environment are you running PnP PowerShell on?
- [x] Windows
- [ ] Linux
- [ ] MacOS
- [ ] Azure Cloud Shell
- [ ] Azure Functions
- [ ] Other : please specify

**Comment by erwinvanhunen** [comment]
This is issue is really hard for us to debug as we do not have access to a GCC tenant (and we will not get access either, as most of us are not living/working in the US, are not US citizens, nor do we work for Microsoft) so debugging this will be a trial and error process.

For my understanding: does a simple cmdlet like 'Get-PnPList' work?

From the parameters you're using I see you're using ACS auth (which is regarded as 'legacy' but available for backwards compatiblity). Could you consider moving over to clientid/certificate based auth?

We will make requests to for instance the Graph behind the scenes in the provisioning engine, and we have been told that the Graph and GCC tenants is a complex story, and using ACS we cannot acquire a Graph access token.

**Comment by steve-boyle** [comment]
In my experience with GCC High, only -UseWebLogin with a user that has appropriate rights works with PnP.  With a certificate style login, the required Sharepoint permissions are missing in GCC High.  I also tried to use Get-Credential and pass that into PnP with similar results to you, didn't work.  I ended up reimplementing my process in Flow, which allows me to use a user with appropriate permissions and also run on a schedule without human interaction.

**Comment by erwinvanhunen** [comment]
Hi. We just merged a change and increased the number of environments we support specifically. Now you can specify a USGovernmentHigh value for -AzureEnvironment (besides USGovernment and USGovernmentDoD). This changes how we acquire tokens behind the scenes (different endpoints) and also how we make calls to the Graph (also different endpoints). It would be great if you could test this with the release that will be available tomorrow as version 1.3.8-nightly (we cannot test it ourselves unfortunately).

**Comment by erwinvanhunen** [comment]
And another update. We have made -many- fixes today and have had confirmed that access to a GCC tenant works now. SharePoint calls (CSOM) and Graph calls have been confirmed to work.

These changes will be availabe in the 1.3.9-nightly build:

```powershell
Update-Module -Name PnP.PowerShell -RequiredVersion 1.3.9-nightly -AllowPrerelease
```

You are required to register your own Azure AD App to access your environment through PnP PowerShell. We currently only support -Interactive and certificate based authentication. Calls to SharePoint work, calls to Teams and calls to Microsoft365 groups have been confirmed to work.

When you register the app the Azure AD you cannot select the appropriate SharePoint permissions, this is a limitation of GCC Tenants. You can add the permissions manually to the manifest though. See below for more info on that. 

You can also register app for both delegate and app only use through PnP PowerShell (we have had this tested)

```powershell
Register-PnPAzureADApp -ApplicationName "yourapplication" -Tenant [yourtenant].onmicrosoft.com -Interactive -AzureEnvironment [USGovernment|USGovernmentHigh|USGovernmentDoD] -Scopes SPO.AllSites.FullControl, SPO.Sites.FullControl.All, MSGraph.Group.ReadWrite.All
```

in the list of scopes: SPO.AllSites.FullControl is a delegate permission for the SharePoint whereas SPO.Sites.FullControl.All an app-only permission is.

You can add these permissions manually too in the manifest of the app in Azure AD. 

Locate the 'requiredResourceAccess' section and add or modify the existing entries. See the example below. Notice, this an example, do not copy and paste this straight away.

```json
"requiredResourceAccess": [
{
    "resourceAppId": "00000003-0000-0ff1-ce00-000000000000",
    "resourceAccess": [
		{
			"id": "56680e0d-d2a3-4ae1-80d8-3c4f2100e3d0",
			"type": "Scope"
		}
      ]
}
```

Use the following values:

The resourceAppId for SharePoint = "00000003-0000-0ff1-ce00-000000000000" 

Permission | Permission type | Id | Type
| -------| ----------- | ------ | ----- |
| Sites.FullControl.All | Application | 678536fe-1083-478a-9c59-b99265e6b0d3 | Role |
| AllSites.FullControl | Delegate | 56680e0d-d2a3-4ae1-80d8-3c4f2100e3d0 | Scope |

After all is setup correctly you can login with

```powershell
Connect-PnPOnline -Url [yoursite] -Interactive -ClientId [clientid of the app] -Tenant [yourtenant].onmicrosoft.com -AzureGovernment USGovernmentHigh
```
